### ISSUE TITLE
Allow for static ERGM resimulation for duration = 1 models

### ISSUE THREAD

**Issue by smjenness** [statnet/EpiModel#528]
Following up on #525. 

Need to allow for static ERGM resimulation here:
https://github.com/statnet/EpiModel/blob/master/R/net.mod.simnet.R#L87-L126

But also, redo network initialization, check network stats, and check reconstruction of networkDynamic object from panel ERGM sims.

**Comment by ThomasKraft** [comment]
@smjenness Just trying to get a handle on when this implementation might be ready since I'll be using for the last step of an ongoing project -- am I correct that it should be a couple of months as part of Intermediate Issues? I'm not meaning to be pushy!

**Comment by smjenness** [comment]
It may be a couple weeks

**Comment by smjenness** [comment]
@ThomasKraft : sorry for the delay on this. It took quite a bit of work on the underlying infrastructure for network resimulation. This is now ready on #546, which will be merged into master shortly. 

Here's a working example:
```
# Your original working example of a very high density network
num = 20
toy <- network_initialize(num)

ft <- netest(toy,  formation = ~edges ,
             target.stats = 100,
             coef.diss = dissolution_coefs(dissolution = ~offset(edges),
                                           duration = 1),
             edapprox = TRUE)
param <- param.net(inf.prob = 0.2, act.rate = 1, rec.rate = 1/3)
init <- init.net(i.num = 10, r.num = 0)
control <- control.net(type = "SIR", nsteps = 25, nsims = 2, ncores = 1,
                       tergmLite = FALSE,  resimulate.network = FALSE, verbose = FALSE)
simz <- netsim(ft, param, init, control)

# get_network converts stored cross-sectional network list into a networkDynamic object
net <- get_network(simz, sim = 2)
net

# The timed edgelist now shows dyad spells of heterogeneous duration that span time steps 
nwd <- as.data.frame(net)
head(nwd, 25)

# You can also pull out the network list directly and work with that
nwl <- get_network(simz, sim = 2, ergm.create.nd = FALSE)

# network structure on first, second, third... time steps
nwl[[1]]
nwl[[2]]
nwl[[3]]
```

**Comment by smjenness** [comment]
There are some more variations here, as I used these for unit testing: https://github.com/statnet/EpiModel/blob/resim-ergm/inst/test-static-ergm.R.

**Comment by ThomasKraft** [comment]
@smjenness Thank you very much for the hard work on this! I loaded the `resim-ergm` branch earlier today but haven't had a chance to run any tests yet. Will do so with more extensive examples tomorrow.

**Comment by ThomasKraft** [comment]
Quick clarification: 

```
# You can also pull out the network list directly and work with that
nwl <- get_network(simz, sim = 2, ergm.create.nd = FALSE)
```

The above is the same as looking at the network list outputted by netsim, correct? (as in no advantage to using `get_network` in that example)? E.g. `identical(nwl, simz$network$sim2)` == T

Also, I noticed in the example that you set `resimulate.network == F`. I assume that as long as demographic or other processes are impacting the network structure that it would make sense to set this parameter to TRUE, even in the case of static ERGMs?

**Comment by smjenness** [comment]
You should use `get_network` rather than `$` to access the network. There is no guarantee that the network objects will be consistently accessible with `$`. 

Yes, you can do `resimulate.network = TRUE` or `FALSE` here.

**Comment by ThomasKraft** [comment]
Thanks! Could you also clarify, from your code:

```
# get_network converts stored cross-sectional network list into a networkDynamic object
net <- get_network(simz, sim = 2)
```

As far as I can tell, the resulting `net` object does not include "testatus" or any dynamic variable indicating infection status. I've also tried looking into the individual network panel objects and don't see that info there. Would reconstructing this info from the transmat be the only option?

**Comment by smjenness** [comment]
I have a solution in the works...

**Comment by ThomasKraft** [comment]
In case it is helpful, here is my modification to `nwupdate.net` to deal with the issue in an SEIR context:

```
nwupdateSEIR.net <- function(dat, at) {
  
  ## Attributes
  type <- get_control(dat, "type", override.null.error = TRUE)
  tergmLite <- get_control(dat, "tergmLite")
  status <- get_attr(dat, "status")
  infTime <- get_attr(dat, "infTime")
  active <- get_attr(dat, "active")
  entrTime <- get_attr(dat, "entrTime")
  exitTime <- get_attr(dat, "exitTime")
  isTERGM <- get_control(dat, "isTERGM")
  
  # statOnNw <- "status" %in% dat$temp$nwterms
  resimulate.network <- get_control(dat, "resimulate.network")
  
  ## Vital Dynamics
  arrivals <- which(active == 1 & entrTime == at)
  departures <- which(active == 0 & exitTime == at)
  
  nArrivals <- length(arrivals)
  if (nArrivals > 0) {
    
    ## Arrivals
    nwterms <- dat$temp$nwterms
    if (!is.null(nwterms)) {
      curr.tab <- get_attr_prop(dat, nwterms)
      dat <- auto_update_attr(dat, arrivals, curr.tab)
    }
    if (length(unique(sapply(dat$attr, length))) != 1) {
      stop("Attribute list of unequal length. Check arrivals.net module.\n",
           print(cbind(sapply(get_attr_list(dat), length))))
    }
    if (tergmLite == FALSE) {
      dat$nw[[1]] <- add.vertices(dat$nw[[1]], nv = nArrivals)
      if (isTERGM == TRUE) {
        dat$nw[[1]] <- activate.vertices(dat$nw[[1]], 
                                         onset = at, terminus = Inf, v = arrivals)
        dat$nw[[1]] <- activate.vertex.attribute(dat$nw[[1]], 
                                                 prefix = "testatus", value = status[arrivals], 
                                                 onset = at, terminus = Inf, v = arrivals)
      }
    }
    if (tergmLite == TRUE) {
      dat$el[[1]] <- add_vertices(dat$el[[1]], nv = nArrivals)
    }
    
  }
  
  
  ## Departures
  if (length(departures) > 0) {
    if (tergmLite == FALSE) {
      if (isTERGM == TRUE) {
        dat$nw[[1]] <- deactivate.vertices(dat$nw[[1]], 
                                           onset = at, terminus = Inf, v = departures, 
                                           deactivate.edges = TRUE)
      }
      else {
        dat$nw[[1]] <- delete.vertices(dat$nw[[1]], vid = departures)
        dat <- delete_attr(dat, departures)
      }
    }
    if (tergmLite == TRUE) {
      dat <- delete_attr(dat, departures)
      dat$el[[1]] <- delete_vertices(dat$el[[1]], departures)
      if (dat$control$tergmLite.track.duration) {
        dat$p[[1]]$state$nw0 %n% "lasttoggle" <- delete_vertices(dat$p[[1]]$state$nw0 %n% 
                                                                   "lasttoggle", departures)
      }
    }
  }
  

  ## Infection
  if (tergmLite == FALSE) {
    idsNewInf <- which(status == "e" & infTime == at)
    if (length(idsNewInf) > 0) {
      if(isTERGM == TRUE){
        dat$nw[[1]] <- activate.vertex.attribute(dat$nw[[1]], prefix = "testatus",
                                               value = "e", onset = at,
                                               terminus = Inf, v = idsNewInf)
      } 
      # else{
      #   dat$nw[[1]] <- set_vertex_attribute(dat$nw[[1]], attrname = "status", value = "e", v = idsNewInf)
      # }
    }
  }
  
  ## Move from "e" to "i" state
  if (tergmLite == FALSE) {
      attr.status <- which(status == "i")
      nw.status <- which(get_vertex_attribute(dat$nw[[1]], "status") == "i")
      idsI <- setdiff(attr.status, nw.status)
      if (length(idsI) > 0) {
        if(isTERGM == TRUE){
          dat$nw[[1]] <- activate.vertex.attribute(dat$nw[[1]], prefix = "testatus",
                                                   value = "i", onset = at,
                                                   terminus = Inf, v = idsI)
        } 
        # else{
        #   dat$nw[[1]] <- set_vertex_attribute(dat$nw[[1]], attrname = "status", value = "i", v = idsI)
        # }
      }
  }
  
  ## Recovery
  if (tergmLite == FALSE) {
    attr.status <- which(status == "r")
    nw.status <- which(get_vertex_attribute(dat$nw[[1]], "status") == "r")
    idsRec <- setdiff(attr.status, nw.status)
    if (length(idsRec) > 0) {
      if(isTERGM == TRUE){
        dat$nw[[1]] <- activate.vertex.attribute(dat$nw[[1]], prefix = "testatus",
                                                 value = "r", onset = at,
                                                 terminus = Inf, v = idsRec)
      } 
      # else{
      #   dat$nw[[1]] <- set_vertex_attribute(dat$nw[[1]], attrname = "status", value = "r", v = idsRec)
      # }
    }
  }
  
  ## Death
  if (tergmLite == FALSE) {
    attr.status <- which(status == "d")
    nw.status <- which(get_vertex_attribute(dat$nw[[1]], "status") == "d")
    idsDead <- setdiff(attr.status, nw.status)
    if (length(idsDead) > 0) {
      if(isTERGM == TRUE){
        dat$nw[[1]] <- activate.vertex.attribute(dat$nw[[1]], prefix = "testatus",
                                                 value = "d", onset = at,
                                                 terminus = Inf, v = idsDead)
      } 
      # else{
      #   dat$nw[[1]] <- set_vertex_attribute(dat$nw[[1]], attrname = "status", value = "d", v = idsDead)
      # }
    }
  }
  
  dat$nw[[1]] <- set_vertex_attribute(dat$nw[[1]], attrname = "status", value = get_attr(dat, "status"))
  
  ## Copy static attributes to network object
  if (tergmLite == FALSE & resimulate.network == TRUE) {
    dat <- copy_datattr_to_nwattr(dat)
  }
  if (tergmLite == FALSE & isTERGM == TRUE) {
    dat$nw[[1]] <- activate.vertex.attribute(dat$nw[[1]], 
                                             prefix = "testatus", value = status, onset = at, 
                                             terminus = Inf)
  }
  if (tergmLite == FALSE & isTERGM == FALSE & resimulate.network == 
      TRUE) {
    dat$temp$nw_list[[at]] <- dat$nw[[1]]
  }
  
  # Attribute consistency checks
  # if (tergmLite == FALSE) {
  #   tst <- get.vertex.attribute.active(dat$nw[[1]], "testatus", at = at)
  #   if (any(is.na(tst))) {
  #     stop("Error in nwupdate.net: NA's in testatus attribute.\n")
  #   }
  #   if (statOnNw == TRUE) {
  #     fstat <- get_vertex_attribute(dat$nw[[1]], "status")
  #     if (!identical(tst, fstat)) {
  #       stop("Error in nwupdate.net: mismatch between status and testatus attribute.\n")
  #     }
  #   }
  # }
  
  ## Output
  return(dat)
}
```

**Comment by ThomasKraft** [comment]
I'm also experiencing some odd behavior when examining the network dataframe within modules. When `at == 2`, `as.data.frame(dat$nw)` returns a data frame with all columns as you would expect for a networkDynamic object (e.g. onset, censor columns, etc.). But when `at > 2`, `as.data.frame(dat$nw)` returns a dataframe with just 2 columns called ".tail" and ".head" (are the '.' characters intended for these headers?).

All this suggest to me that in the static case the network is still being initialized as a networkDynamic object before being changed -- is that the intended behavior?

**Comment by smjenness** [comment]
Can you try installing from the `ergm-testatus` branch and see if that works for you? It now saves out status to the cross-sectional network panels and creates a temporally extended attribute `status.active` when extracting the network with `get_network`. It should also address the issue directly above about consistency with network class objects across the time steps.

```r
remotes::install_github("statnet/EpiModel@ergm-testatus")
```

I cannot merge this branch into master right now due to some unrelated R CMD Check errors...

**Comment by ThomasKraft** [comment]
The status implementation is working well. Unfortunately this introduced a new error into the following: 

```
# get_network converts stored cross-sectional network list into a networkDynamic object
net2 <- get_network(sim2, sim = 1)

Error in networkDynamic(network.list = x$network[[sim]], start = 1, base.net = x$network[[sim]][[1]],  : 
  vertex.pid must be specified when network sizes of network.list are different
```
In this case my best guess is that this is caused by individuals dying and departing from the model. Would it be best to specify the vertex.pid's directly from the saved networks somehow?

**Comment by smjenness** [comment]
I imagined that something like that might happen because I didn't test against open population models. That is because it just gets really tricky to develop a general implementation here. For one-time step ties, we have to use the Statnet machinery for static networks (ergm models and network objects), rather than the machinery specifically developed for open, dynamic populations (tergm models and networkDynamic objects). networkDynamic has the capability of "deactivating vertices" (keeping them around but not letting them form new edges) whereas one must delete vertices for network objects. The epidemic simulation should still work fine, but some of the individual-level data extraction and analysis may not be possible to support generally. You might need to save and track any individual-level data you need in a bespoke manner for your specific research project (e.g., with saving attributes out over time in a matrix).

**Comment by ThomasKraft** [comment]
Ok that's great to know. Can you verify that it wouldn't be possible in the `get_network` call to simply specify `vertex.pid = "vertex.names"`.  As far as I can tell, the "vertex.names" attribute of the static network panel stays consistent throughout (e.g. removing node 10 does not change the vertex.names of the other nodes). Would that be a solution to this particular instance?

**Comment by smjenness** [comment]
I am surprised that deleting (or adding) vertices does not mess up the vertex.names in the static networks. If you can test and verify that, it should be ok to use as the pid.

**Comment by ThomasKraft** [comment]
Right I should probably be very careful then. Would it perhaps be safer to assign a new vertex attribute at the beginning of unique ids (uids?)? Given that departing nodes are deleted along with their attributes in the static case, this seems safe?

**Comment by smjenness** [comment]
Yes, you could use `attr$unique_id` which is initialized with `append_core_attr` in recent versions of EpiModel. You might need to set it on the network list each time step (`e.g., dat$temp$nw_list <- set_vertex_attribute(dat$temp$nw_list, ....)`) so it gets saved for later extraction with `get_network`. Sorry this is so clunky. If I had more time I would develop a cleaner solution.

**Comment by ThomasKraft** [comment]
Thanks, that is super helpful. I'll see if I can find a way to tell `get_network` to set `vertex.pid = 'unique_id'` for the networkDynamic call. I really do appreciate the help, and apologize for taking up so much of your time with this edge case problem!

**Comment by smjenness** [comment]
No worries - I am glad I fixed the problem you identified with single-time-step contacts. Even if the individual level data may be clunky to work with, the epi sims themselves should be good now.

**Comment by ThomasKraft** [comment]
So sorry to bother you again @smjenness, but could you please add the following argument:

```
constraints = nwparam$constraints
```

 to the `resim_nets` module for the static ERGM case at: https://github.com/statnet/EpiModel/blob/fe397827a574756ea51c54268f594d112795d1d5/R/net.mod.simnet.R#L175-L181?

It would probably be a good idea to keep the constraints parameter in any associated `simulate` calls for the static ERGM case as well. Especially with the new and amazing constraints functionality in v4.0!

**Comment by smjenness** [comment]
@ThomasKraft : added on `master` branch now with 722f26837f21c032ae5a5f9e9c54d4b9c8a5310e.

**Comment by ThomasKraft** [comment]
@smjenness : Just to be 100% sure, am I correct that `ergm-testatus` has been merged with `master` and now all static ergm functionality is supported on `master`? Everything I've checked indicates that this is the case.

**Comment by smjenness** [comment]
yes, it's merged. I am deleting old branches now...

**Comment by ThomasKraft** [comment]
One more (hopefully final!) question for the static case. I am doing some manual dissolution of ties in a custom module, which apparently causes problems at the following step when the network object (`dat$nw[[1]]`) is used in the `basis` argument of `simulate.ergm`:

https://github.com/statnet/EpiModel/blob/722f26837f21c032ae5a5f9e9c54d4b9c8a5310e/R/net.mod.simnet.R#L177-L184

Apparently, modifying `dat$nw[[1]]` and then resimulating from it in this way causes the number of edges to reflect the modified network object, rather than coefficients in the formula. It's a bit unclear to me what the purpose of the `basis` argument is here ("to override the LHS of the formula"?), but it seems that a valid workaround would be to instead specify `basis = dat$temp$nw_list[[1]]` in the linked code above to circumvent the problem. Could you verify that this would be valid, and not a problem to use the LHS of the original network object in this way throughout a simulation?

**Comment by ThomasKraft** [comment]
It would appear that this is actually a constraints problem that arises when attributes dynamically change in the network. Here is a reprex to demonstrate:

```
num=20
toy <- network_initialize(n=num, directed=F)
toy <- set_vertex_attribute(toy, "com_id", c(rep("1", num/2),rep("2", num/2)))

ft <- netest(toy,
                    formation = ~edges,
                    constraints =~blocks("com_id", levels2=~!diag(TRUE, length(unique(.nw%v%"com_id")))),
                    target.stats = c(70),
                    coef.diss = dissolution_coefs(dissolution = ~offset(edges), duration = 1), edapprox = T)

test <- simulate(object = ft$formation,
                 basis = toy,
                 coef = ft$coef.form,
                 constraints = ft$constraints,
                 # control = mcmc.control.ergm,
                 dynamic = FALSE,
                 nsim = 1)

# Produces expected amount of edges
simulate(object = ft$formation,
         basis = test,
         coef = ft$coef.form,
         constraints = ft$constraints,
         # control = mcmc.control.ergm,
         dynamic = FALSE,
         nsim = 1)


# case in which the attribute governing the constraint changes
# Does NOT produce expected amount of edges
test2 <- test
test2 %v% "com_id" <- sample(c("1", "2"), 20, replace=T)

prob <-simulate(object = ft$formation,
         basis = test2,
         coef = ft$coef.form,
         constraints = ft$constraints,
         # control = mcmc.control.ergm,
         dynamic = FALSE,
         nsim = 1)

summary(prob~nodematch("com_id"))  # can see that constraint has not been upheld after attribute change
```

I don't understand why `simulate` here is not making use of the attributes in the current network. My best guess is that it has something to do with the LHS of the underlying "ft" object. In the static case my goal is to resimulate the network based on the original `netest` parameters but using a modified version of the network as basis. Perhaps this is more of a question for @chad-klumb ?

**Comment by ThomasKraft** [comment]
I should also note that there seems to be a rather disappointing workaround for this, which is to use a -Inf coefficient instead of the proper constraint, e.g.:

```
ft2 <- netest(toy,
             formation = ~edges +
               offset(absdiff(~as.integer(com_id))),
             coef.form = c(-Inf),
             target.stats = c(70),
             coef.diss = dissolution_coefs(dissolution = ~offset(edges), duration = 1), edapprox = T)

test <- simulate(object = ft2$formation,
         basis = toy,
         coef = ft2$coef.form,
         constraints = ft2$constraints,
         # control = mcmc.control.ergm,
         dynamic = FALSE,
         nsim = 1)

# DOES still produce expected amount of edges
test2 <- test
test2 %v% "com_id" <- sample(c("1", "2"), 20, replace=T)

simulate(object = ft2$formation,
         basis = test2,
         coef = ft2$coef.form,
         constraints = ft2$constraints,
         dynamic = FALSE,
         nsim = 1)
```

I say disappointing only because this loses all the efficiency perks of the otherwise working constraints infrastructure.

**Comment by smjenness** [comment]
@krivit or @chad-klumb : do you have any thoughts or a potential solution about this issue of attributes referenced in constraints? See only the last two comments in this (long) issue.

**Comment by chad-klumb** [comment]
The `test2` basis does not satisfy the constraint:

```
> summary(test2 ~ nodematch("com_id"))
nodematch.com_id 
              28 
```

**Comment by ThomasKraft** [comment]
In the generating user case ties that do not satisfy the constraint are removed manually prior to the `simulate` call, e.g.:

```
reids <- unique(unlist(sapply(c(1:10), FUN = function(x) get.edgeIDs(x=test, v=x, neighborhood = "combined"))))
test2 <- network::delete.edges(test, e = reids)
```

**Comment by chad-klumb** [comment]
remind me, do you want only ties matching on `com_id`, or only ties not matching on `com_id`?

**Comment by ThomasKraft** [comment]
Only matching on `com_id`, see my edited last comment for the code that ensures the basis is satisfied prior to simulate.

**Comment by chad-klumb** [comment]
How does

```
summary(prob~nodematch("com_id"))  # can see that constraint has not been upheld after attribute change
```

indicate that the constraint has not been upheld?

**Comment by chad-klumb** [comment]
One possible issue here is that you have a small network and you're (stochastically) changing the number of toggleable dyads in resampling the attribute.

If you use

```
toy <- set_vertex_attribute(toy, "com_id", c(rep("1", num/2),rep("2", num/2)))
```

for estimation, you get exactly 10 nodes of each type, while if you resample according to

```
test2 %v% "com_id" <- sample(c("1", "2"), 20, replace=T)
```

you get some random number of each type, with mean 10, but significant CV.

The mean number of edges should not in general be expected to remain the same if the number of free dyads is changed but the generative model is unchanged.

**Comment by chad-klumb** [comment]
Here's an example with a larger network.

```
require(EpiModel)
num <- 2000
toy <- network_initialize(n=num, directed=FALSE)
toy <- set_vertex_attribute(toy, "com_id", c(rep("1", num/2),rep("2", num/2)))

ft <- netest(toy,
             formation = ~edges,
             constraints =~blocks("com_id", levels2=~!diag(TRUE, length(unique(.nw%v%"com_id")))),
             target.stats = c(7000),
             coef.diss = dissolution_coefs(dissolution = ~offset(edges), duration = 1), edapprox = T)

ft$fit$newnetwork

test <- simulate(object = ft$formation,
                 basis = toy,
                 coef = ft$coef.form,
                 constraints = ft$constraints,
                 control = list(MCMC.burnin=1e5,MCMC.interval=1e5),
                 dynamic = FALSE,
                 monitor = ~nodematch(~com_id),
                 output = "stats",
                 nsim = 10)

test[,1:2]

test2 <- toy
test2 %v% "com_id" <- sample(c("1", "2"), 2000, replace=T)

prob <-simulate(object = ft$formation,
                basis = test2,
                coef = ft$coef.form,
                constraints = ft$constraints,
                control = list(MCMC.burnin=1e5,MCMC.interval=1e5),
                dynamic = FALSE,
                monitor = ~nodematch(~com_id),
                output = "stats",
                nsim = 10)

prob[,1:2]
```

I get

```
> test[,1:2]
      edges nodematch.com_id
 [1,]  7021             7021
 [2,]  7028             7028
 [3,]  7126             7126
 [4,]  7072             7072
 [5,]  6820             6820
 [6,]  7080             7080
 [7,]  7099             7099
 [8,]  7047             7047
 [9,]  6753             6753
[10,]  7026             7026
```

and

```
> prob[,1:2]
      edges nodematch.com_id
 [1,]  7188             7188
 [2,]  7049             7049
 [3,]  7080             7080
 [4,]  6970             6970
 [5,]  6815             6815
 [6,]  7070             7070
 [7,]  6807             6807
 [8,]  6954             6954
 [9,]  6972             6972
[10,]  6995             6995
```

so we get about the same number of edges (since the relative change in the number of free dyads is much smaller), and the constraint is respected.

**Comment by ThomasKraft** [comment]
Thanks @chad-klumb, yes I was using a larger network when doing this for myself and had learned through trial and error that issue with the small networks. 

But I am now realizing that this whole issue has been a red herring! When you ran the code above, you included `control = list(MCMC.burnin=1e5,MCMC.interval=1e5)`, which is something that I have been specifying in my test cases for EpiModel in `control.net` as `set.control.ergm=control.simulate.ergm(MCMC.burnin=2e7, MCMC.interval=2000)`. It turns out that the bug all along was that this argument was not being passed along to the call to `simulate` for the static ergm (duration = 1) case. 

For example, see here: https://github.com/statnet/EpiModel/blob/722f26837f21c032ae5a5f9e9c54d4b9c8a5310e/R/net.mod.simnet.R#L177-L184

Thus, the following needs to be added to `resim_nets` in order to get the control parameters from `control.net` to get used:
```
set.control.ergm <- get_control(dat, "set.control.ergm")
set.control.ergm <- set.control.ergm[-which(names(set.control.ergm) == "MCMC.scale")] # this argument not included for control.simulate.ergm.formula 
```

So I was getting problematic results when running on a larger network because the control parameter wasn't working and was too small to reconfigure the network. Does that make sense?

**Comment by smjenness** [comment]
@chad-klumb : do you agree with this approach to adding set.control.ergm in that spot?

We need control parameters that work for both simulating from the ergm object (like here: https://github.com/statnet/EpiModel/blob/master/R/net.mod.simnet.R#L72) and also simulating from a formula (like here: https://github.com/statnet/EpiModel/blob/master/R/net.mod.simnet.R#L181). 

Thanks

**Comment by chad-klumb** [comment]
IMO EpiModel `netsim` should consistently do formula-style simulation.  It should be possible to convert the simulation from the `ergm` object to a formula simulation.  In doing so, we could combine the `set.control.ergm` and `mcmc.control.ergm` arguments to `control.net` into a single `set.control.ergm` argument of class `control.simulate.formula`.  This would introduce a slight backwards incompatibility for controls (since it would change the class of `set.control.ergm`), but overall would be an improvement in simplicity and consistency.

**Comment by smjenness** [comment]
I like that approach. We can discuss further on our call later today before this gets implemented, because I have a couple related questions to this section of the code that I'd like to discuss with you first...

**Comment by ThomasKraft** [comment]
Further testing of the static ergm case is also indicating memory management problems, similar to what existed previously before some of the fixes to `trim_env()` in https://github.com/statnet/tergmLite/issues/48#issue-798521359. The behavior I'm observing is that memory usage monitored through htop of each process running in parallel continues increasing dramatically over the course of a run (with a network of ~9000 nodes, up to 65 GB/process or so). This was the case previously for the non-static case before the issue linked above -- is there any reason to believe that `trim_env` would be functioning differently in the static case? I could open a new issue for this, but seems tied to the duration = 1 resimulation implementation.

**Comment by martinamorris** [comment]
IIRC the main benefit of simulating from the fit object rather than the formula was that the fit object sim allows you to start with the `newnetwork`, and avoid the burnin from an empty network.  But if `simulate.formula` has a `basis` argument where you can pass the `newnetwork` that might retain the advantage.

**Comment by ThomasKraft** [comment]
I think the memory issue in the static ergm case may result from the environment in the call to `.simulate_formula.network`: https://github.com/statnet/ergm/blob/e65c4bf89f7a065149dee9c86eef8aad1601431a/R/simulate.ergm.R#L268-L280

Simulating from a formula as linked above does not make use of `trim_env()` as is done for the reference and constraints in `simulate.ergm_model` here: https://github.com/statnet/ergm/blob/e65c4bf89f7a065149dee9c86eef8aad1601431a/R/simulate.ergm.R#L358-L370

Or for tergm here: https://github.com/statnet/tergm/blob/398c8a103b5c121c585887ae1e6db246e6bdadb8/R/simulate.stergm.R#L226-L227

If I'm remembering correctly, when this happened previously with constraints @krivit fixed in EpiModel directly?
